// stillat-blade-parser/out/parser/stringSplitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringSplitter = void 0;
const isStartOfString_1 = require("./scanners/isStartOfString");
const skipToEndOfString_1 = require("./scanners/skipToEndOfString");
class StringSplitter {
    constructor() {
        this.chars = [];
        this.cur = null;
        this.prev = null;
        this.next = null;
        this.currentIndex = 0;
        this.inputLen = 0;
        this.currentContent = [];
    }
    advance(count) {
        for (let i = 0; i < count; i++) {
            this.currentIndex++;
            this.checkCurrentOffsets();
        }
    }
    encounteredFailure() {
        return;
    }
    updateIndex(index) {
        this.currentIndex = index;
    }
    inputLength() {
        return this.inputLen;
    }
    getCurrentIndex() {
        return this.currentIndex;
    }
    incrementIndex() {
        this.currentIndex += 1;
    }
    getCurrent() {
        return this.cur;
    }
    getNext() {
        return this.next;
    }
    getPrev() {
        return this.prev;
    }
    checkCurrentOffsets() {
        this.cur = this.chars[this.currentIndex];
        this.prev = null;
        this.next = null;
        if (this.currentIndex > 0) {
            this.prev = this.chars[this.currentIndex - 1];
        }
        if ((this.currentIndex + 1) < this.inputLen) {
            this.next = this.chars[this.currentIndex + 1];
        }
    }
    pushChar(value) {
        this.currentContent.push(value);
    }
    getChar(index) {
        return this.chars[index];
    }
    getSeedOffset() {
        return 0;
    }
    getContentSubstring(from, length) {
        return '';
    }
    static fromText(value) {
        const splitter = new StringSplitter();
        return splitter.split(value);
    }
    split(value) {
        this.chars = value.split('');
        this.inputLen = this.chars.length;
        const returnStrings = [];
        for (this.currentIndex = 0; this.currentIndex < this.inputLen; this.currentIndex += 1) {
            this.checkCurrentOffsets();
            if ((0, isStartOfString_1.isStartOfString)(this.cur)) {
                const results = (0, skipToEndOfString_1.skipToEndOfStringTraced)(this);
                this.currentContent.push(this.cur);
                this.currentContent = this.currentContent.concat(results.value.split(''));
                this.currentContent.push(this.cur);
                this.currentIndex = results.endedOn;
                if (this.currentIndex == this.inputLen - 1) {
                    returnStrings.push(this.currentContent.join(''));
                    break;
                }
                continue;
            }
            if (this.cur == ',' || this.next == null) {
                if (this.cur != ',') {
                    this.currentContent.push(this.cur);
                }
                returnStrings.push(this.currentContent.join(''));
                this.currentContent = [];
                continue;
            }
            this.currentContent.push(this.cur);
        }
        return returnStrings;
    }
}
exports.StringSplitter = StringSplitter;
