// stillat-blade-parser/out/errors/bladeError.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BladeError = exports.ErrrorLevel = void 0;
var ErrrorLevel;
(function (ErrrorLevel) {
    ErrrorLevel[ErrrorLevel["Error"] = 0] = "Error";
    ErrrorLevel[ErrrorLevel["Warning"] = 1] = "Warning";
})(ErrrorLevel = exports.ErrrorLevel || (exports.ErrrorLevel = {}));
class BladeError {
    constructor() {
        this.node = null;
        this.errorCode = '';
        this.message = '';
        this.level = ErrrorLevel.Error;
        this.range = null;
    }
    hash() {
        var _a, _b, _c, _d;
        let positionSlug = '';
        if (this.node != null) {
            positionSlug = ((_b = (_a = this.node.endPosition) === null || _a === void 0 ? void 0 : _a.offset) !== null && _b !== void 0 ? _b : 0).toString() + "|" +
                ((_d = (_c = this.node.startPosition) === null || _c === void 0 ? void 0 : _c.offset) !== null && _d !== void 0 ? _d : 0).toString() + "|";
        }
        return positionSlug + "|" + this.errorCode;
    }
    static makeSyntaxError(errorCode, node, message, level) {
        if (level == null) {
            level = ErrrorLevel.Error;
        }
        const error = new BladeError();
        error.errorCode = errorCode;
        error.node = node;
        error.message = message;
        error.level = level;
        return error;
    }
}
exports.BladeError = BladeError;
