// stillat-blade-parser/out/document/documentErrors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentErrors = void 0;
class DocumentErrors {
    constructor(doc) {
        this.doc = doc;
    }
    hasStructureErrors() {
        return this.doc.getParser().getStructureErrors().length > 0;
    }
    getFirstStructureError() {
        return this.doc.getParser().getStructureErrors()[0];
    }
    hasAny() {
        return this.all().length > 0;
    }
    all() {
        const errorHashes = [], errors = [];
        this.doc.getAllNodes().forEach((node) => {
            node.getErrors().forEach((error) => {
                if (errorHashes.includes(error.hash()) == false) {
                    errorHashes.push(error.hash());
                    errors.push(error);
                }
            });
        });
        this.doc.getParser().getErrors().forEach((error) => {
            if (errorHashes.includes(error.hash()) == false) {
                errorHashes.push(error.hash());
                errors.push(error);
            }
        });
        return errors;
    }
}
exports.DocumentErrors = DocumentErrors;
