// stillat-blade-parser/out/nodes/position.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Position = void 0;
class Position {
    constructor() {
        this.index = -1;
        this.offset = 0;
        this.line = 0;
        this.char = 0;
    }
    shiftRight(shiftOffset) {
        const position = new Position();
        position.offset = this.offset + shiftOffset;
        position.index = this.index + shiftOffset;
        position.line = this.line;
        position.char = this.char + shiftOffset;
        return position;
    }
    isBefore(position) {
        if (position == null) {
            return false;
        }
        if (position.line > this.line) {
            return true;
        }
        if (position.line == this.line && this.offset < position.offset) {
            return true;
        }
        return false;
    }
    isWithin(start, end, shiftEnd = 0) {
        if (start == null || end == null) {
            return false;
        }
        if (this.index >= start.index && this.index <= (end.index + shiftEnd)) {
            return true;
        }
        return false;
    }
    isWithinRange(range) {
        if (range == null) {
            return false;
        }
        if (range.start != null && range.end != null) {
            return this.isWithin(range.start, range.end);
        }
        return false;
    }
}
exports.Position = Position;
