// stillat-blade-parser/out/compiler/compilers/forElse.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForElseCompiler = void 0;
class ForElseCompiler {
    constructor() {
        this.loops = 1;
        this.activeLoop = 1;
    }
    getActiveLoop() {
        return this.activeLoop;
    }
    popLoop() {
        this.activeLoop -= 1;
        this.loops -= 1;
    }
    compile(node) {
        const directive = node, content = directive.getInnerContent(), split = this.getVars(content), loopVar = split[0].trim(), tempVar = split[1].trim();
        this.activeLoop = this.loops;
        let result = '<?php $__empty_' + this.activeLoop + ' = true; $__currentLoopData = ' + loopVar + '; $__env->addLoop($__currentLoopData); foreach($__currentLoopData as ' + tempVar + '): $__env->incrementLoopIndices(); $loop = $__env->getLastLoop(); $__empty_' + this.activeLoop + ' = false; ?>';
        this.loops += 1;
        return result;
    }
    getVars(content) {
        const split = content.split(' as ');
        if (split.length == 1) {
            return content.split(' AS ');
        }
        return split;
    }
}
exports.ForElseCompiler = ForElseCompiler;
