// stillat-blade-parser/out/document/printers/stringBuffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringBuffer = void 0;
class StringBuffer {
    constructor(tabSize, activeLevel) {
        this.tabSize = 0;
        this.content = '';
        this.activeIndentLevel = 0;
        this.tabSize = tabSize;
        this.activeIndentLevel = activeLevel;
    }
    appendNoRepeat(text) {
        if (this.content.trimEnd().endsWith(text) == false) {
            return this.append(text);
        }
        return this;
    }
    append(text) {
        this.content += text;
        return this;
    }
    getContents() {
        return this.content;
    }
    newLine() {
        this.content += "\n";
        return this;
    }
    indent() {
        this.content += ' '.repeat(this.tabSize * this.activeIndentLevel);
        return this;
    }
    outdent() {
        let target = this.activeIndentLevel - 1;
        if (target < 1) {
            target = 1;
        }
        this.content += ' '.repeat(this.tabSize * target);
        return this;
    }
}
exports.StringBuffer = StringBuffer;
