// stillat-blade-parser/out/test/testUtils/assertions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assertPaired = exports.assertPosition = exports.assertNotNull = exports.assertNull = exports.assertEchoContent = exports.assertCommentContent = exports.assertLiteralContent = exports.assertInstanceOf = exports.assertCount = void 0;
const assert = require("assert");
function assertCount(count, actual) {
    assert.strictEqual(actual.length, count);
}
exports.assertCount = assertCount;
function assertInstanceOf(expected, instance) {
    assert.strictEqual(instance instanceof expected, true);
}
exports.assertInstanceOf = assertInstanceOf;
function assertLiteralContent(expected, node) {
    const comment = node;
    assert.strictEqual(comment.content, expected);
}
exports.assertLiteralContent = assertLiteralContent;
function assertCommentContent(expected, node) {
    const comment = node;
    assert.strictEqual(comment.innerContent, expected);
}
exports.assertCommentContent = assertCommentContent;
function assertEchoContent(node, content, sourceContent) {
    assert.strictEqual(node.content, content);
    assert.strictEqual(node.sourceContent, sourceContent);
}
exports.assertEchoContent = assertEchoContent;
function assertNull(value) {
    assert.strictEqual(value === null, true);
}
exports.assertNull = assertNull;
function assertNotNull(value) {
    assert.strictEqual(value !== null, true);
}
exports.assertNotNull = assertNotNull;
function assertPosition(position, line, char) {
    assertNotNull(position);
    const checkPos = position;
    assert.strictEqual(checkPos.line, line);
    assert.strictEqual(checkPos.char, char);
}
exports.assertPosition = assertPosition;
function assertPaired(open, close) {
    const openDirective = open, closeDirective = close;
    assertNotNull(openDirective.isClosedBy);
    assertNotNull(closeDirective.isOpenedBy);
    assert.strictEqual(openDirective.isClosedBy, closeDirective);
    assert.strictEqual(closeDirective.isOpenedBy, openDirective);
}
exports.assertPaired = assertPaired;
