// stillat-blade-parser/out/parser/php/phpParserPhpValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhpParserPhpValidator = void 0;
const php_parser_1 = require("php-parser");
class PhpParserPhpValidator {
    constructor() {
        this.lastError = null;
    }
    isValid(template, isEval = true) {
        const parser = new php_parser_1.Engine({
            parser: {
                extractDoc: true,
            },
            ast: {
                withPositions: true,
                withSource: true,
            },
        });
        try {
            if (isEval) {
                parser.parseEval(template);
            }
            else {
                parser.parseCode(template, 'test.php');
            }
        }
        catch (err) {
            if (err instanceof SyntaxError) {
                this.lastError = err;
            }
            return false;
        }
        this.lastError = null;
        return true;
    }
    getLastError() {
        return this.lastError;
    }
}
exports.PhpParserPhpValidator = PhpParserPhpValidator;
